csharp_dotnetcore/17.Proactive-Messages/ProactiveBot.cs

```diff
@@ -0,0 +1,208 @@
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Configuration;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.BotBuilderSamples
+{
+    /// <summary>
+    /// For each interaction from the user, an instance of this class is called.
+    /// This is a Transient lifetime service.  Transient lifetime services are created
+    /// each time they're requested. For each Activity received, a new instance of this
+    /// class is created. Objects that are expensive to construct, or have a lifetime
+    /// beyond the single Turn, should be carefully managed.
+    /// </summary>
+    public class ProactiveBot : IBot
+    {
+        /// <summary>The name of events that signal that a job has completed.</summary>
+        public const string JobCompleteEventName = "jobComplete";
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ProactiveBot"/> class.</summary>
+        /// <param name="accessors">The state accessors for use with the bot.</param>
+        /// <param name="endpointService">The <see cref="EndpointService"/> portion of the <see cref="BotConfiguration"/>.</param>
+        public ProactiveBot(ProactiveAccessors accessors, EndpointService endpointService)
+        {
+            StateAccessors = accessors ?? throw new ArgumentNullException(nameof(accessors));
+
+            // AppId was validated during startup.
+            AppId = endpointService.AppId;
+        }
+
+        /// <summary>Gets the bot's app ID.</summary>
+        private string AppId { get; }
+
+        /// <summary>Gets the state accessors for use with the bot.</summary>
+        private ProactiveAccessors StateAccessors { get; }
+
+        /// <summary>
+        /// Every conversation turn will call this method.
+        /// </summary>
+        /// <param name="turnContext">A <see cref="ITurnContext"/> containing all the data needed
+        /// for processing this conversation turn. </param>
+        /// <param name="cancellationToken">(Optional) A <see cref="CancellationToken"/> that can be used by other objects
+        /// or threads to receive notice of cancellation.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        public async Task OnTurnAsync(ITurnContext turnContext, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            // See https://aka.ms/about-bot-activity-message to learn more about the message and other activity types.
+            if (turnContext.Activity.Type != ActivityTypes.Message)
+            {
+                // Handle non-message activities.
+                await OnSystemActivityAsync(turnContext);
+            }
+            else
+            {
+                // Get the job log from the turn context.
+                JobLog jobLog = await StateAccessors.JobLogData.GetAsync(turnContext, () => new JobLog());
```

Comment:
This is unique to this sample, and could use quite a bit more explication as to what's happening.

Reply:
OnSystemActivity is not a necessary feature here, it was just how I was initially breaking up the event handling from the message handling. I've seen this in some of Jeff's code.